src/app/lib/device/airplay.js

```diff
@@ -1,53 +1,80 @@
 (function (App) {
     'use strict';
 
-    var browser = require('airplay-js').createBrowser(),
+    var Bonjour = require('bonjour'),
+        AirPlayProtocol = require('airplay-protocol');
+
+    var bonjourBrowser = new Bonjour(),
         collection = App.Device.Collection;
 
-    var makeID = function (baseID) {
-        return 'airplay-' + baseID.replace(':', '');
+    var makeID = function (bonjourService) {
+        // strong ID since there can only be 1 bonjour service per host+port
+        var uniqueDeviceId = bonjourService.host + '_' + bonjourService.port;
+        return 'airplay-' + Common.md5(uniqueDeviceId);
     };
 
-    var Airplay = App.Device.Generic.extend({
+    var AirPlay = App.Device.Generic.extend({
         defaults: {
             type: 'airplay',
             typeFamily: 'external'
         },
-        makeID: makeID,
+
+        _makeID: makeID,
+
         initialize: function (attrs) {
             this.device = attrs.device;
-            this.attributes.id = this.makeID(this.device.serverInfo.macAddress || this.device.serverInfo.deviceId || '' + this.device.id);
-            this.attributes.name = this.device.name || this.device.serverInfo.model;
-            this.attributes.address = this.device.info[0];
+            this.attributes.id = this._makeID(this.device.service);
+            this.attributes.name = this.device.service.name;
+            this.attributes.address = this.device.service.host;
         },
+
         play: function (streamModel) {
-            var url = streamModel.attributes.src;
-            this.device.play(url);
+            var self = this;
+            var url = streamModel.get('src');
+            this.device.player.play(url, function (err, res, body) {
+                if (err) {
+                    // FIXME: do something on failure to play
```

Comment:
@xaiki this.stop() would send the airplay command to tell the AirPlay receiver to stop, and yeah that's a good basic start.

For better handling, I think we'd want something like an error message in the Butter interface, or to stop overall playback in the Butter interface.

But I am new to this project and don't know how to do that, so I'll leave those more advanced error handlings to someone else.

I'll add this.stop() for now.